### PR TITLE
Registration with Spotify Omniauth

### PR DESCRIPTION
**Pull request by teresa-m-knowles**
Screen Shot 2019-04-02 at 11 34 12 PM

Styled the nav bar a bit with sass. Not great at it but getting a better understanding of the directories/structure.

**Created a Users table** with: 
--spotify_id: this is the unique identifier that spotify assigns users. It's a string of letters.
--name: Spotify username
-- access_token: Spotify token. It expires every hour and it's used to make requests
-- refresh_token: Used to refresh the access_token. 

Still not sure how the refresh_token and access_token work exactly. I've been reading about it but I think I'll get a better understanding of it later on.

Have not done a Sessions controller yet, for logging in, but I setup the users#create (registration) action so that that user would then be automatically logged in after registration. 

Also created a method in rails_helper to stub the Omniauth registration(I guess also login). 

Finally, **the spotify_id for each user is unique.** So, if a user has already registered in our app, they won't be able to register again by clicking on "Register" because their spotify_id will already be in use. 
We can decide as a group if this is fine or if we want one button that serves both purposes: it registers a user if they're not in our app and it logs them in if they are.

TL;DR: Registration is done and should work. If it breaks, let's look more into the refresh token of Spotify.

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+class UsersController < ApplicationController
+  def create
+    auth = request.env['omniauth.auth']
+    
+    spotify_info = Hash.new
```
Comment (teresa-m-knowles) [inline app/controllers/users_controller.rb:5]: This could probably use strong_params instead

**Review by manojpanta — APPROVED** [review]

File: app/assets/stylesheets/nav.scss [hunk 2]
```diff
@@ -0,0 +1,25 @@
+$spotifygreen: #1DB954;
+
+
+nav.main-nav {
```
Comment (Mackenzie-Frey) [inline app/assets/stylesheets/nav.scss:4]: Let's merge this and change in the future to a sticky nav bar that is collapsible to a hamburger for mobile purposes (bootstrap)

File: app/assets/stylesheets/nav.scss [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  ul {
+    list-style-type: none;
+    text-align: center;
+
+    li {
+      display: inline;
+      a {
+          color: white;
+          padding: 14px;
+          &:hover {
```
Comment (Mackenzie-Frey) [inline app/assets/stylesheets/nav.scss:17]: This is cool, I wonder if there is an easy way to have the hover effect without the underline.
image

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    spotify_info[:access_token] = auth.credentials.token
+    spotify_info[:refresh_token] = auth.credentials.refresh_token
+
+    @user = User.new(spotify_info)
+
+    if @user.save
+      flash[:success] = "You have successfully registered an account"
+      session[:user_id] = @user.id
+    else
+      flash[:error] = @user.errors.full_messages.to_sentence
```
Comment (Mackenzie-Frey) [inline app/controllers/users_controller.rb:17]: full_messages.to_sentece is cool.

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
+class UsersController < ApplicationController
+  def create
```
Comment (Mackenzie-Frey) [inline app/controllers/users_controller.rb:2]: This method is ~20 lines. Thinking there is a way to refactor in the future, but I'd prefer to move forward due to time constraints. I'll create a Refactor card on the project board and add this method to it.

File: app/helpers/users_helper.rb [hunk 6]
```diff
@@ -0,0 +1,2 @@
+module UsersHelper
```
Comment (Mackenzie-Frey) [inline app/helpers/users_helper.rb:1]: What does this file do?

File: spec/features/user_can_register_with_spotify_spec.rb [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+            visit '/'
+            click_button "Register with Spotify"
+          end
+
+          it 'should redirect me to the dashboard' do
+            expect(current_path).to eq('/')
+          end
+
+          it 'the navigation bar should now include Logout' do
+
```
Comment (Mackenzie-Frey) [inline spec/features/user_can_register_with_spotify_spec.rb:25]: I really like how many `it` blocks you have. It makes it very easy to walk through the test and understand what's going on. I do think the spaces after the `it` line are unnecessary stylistically (lines 25, 41, 39).

File: spec/features/user_can_register_with_spotify_spec.rb [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+            expect(User.count).to eq(1)
+            expect(User.last.name).to eq('Fake Name')
+          end
+        end
+
+        describe 'if my spotify account is already registered in the app' do
+          before :each do
+            stub_oauth_registration
+            Rails.application.env_config['omniauth.auth'] = OmniAuth.config.mock_auth[:spotify]
+            # This is the same spotify_id from the stubbed_omniauth
```
Comment (Mackenzie-Frey) [inline spec/features/user_can_register_with_spotify_spec.rb:49]: Perfect scenario for a comment.

File: spec/features/user_can_register_with_spotify_spec.rb [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+            create(:user, spotify_id: "fakeId")
+
+          end
+          it 'I see an error message' do
+            visit '/'
+            click_button "Register with Spotify"
+
+            expect(page).to have_content("Spotify has already been taken")
```
Comment (Mackenzie-Frey) [inline spec/features/user_can_register_with_spotify_spec.rb:57]: I'd love it if we made this error message more descriptive. 
ie 'This Spotify account has already been registered with DJ Spoti, please click on Log In.'

File: spec/models/user_spec.rb [hunk 10]
```diff
@@ -0,0 +1,13 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  describe 'validations' do
```
Comment (Mackenzie-Frey) [inline spec/models/user_spec.rb:4]: I appreciate you including the model validations.

**Review by Mackenzie-Frey — APPROVED** [review]
Great job Teresa! I checked out on localhost and all looks good to me. Looks like we need redirect the user to '/dashboard' once they are logged in (I added this to the Log In project board card).
 Also, I added a new card on the project board for styling error/flash messages.

File: app/assets/stylesheets/nav.scss (hunk 3 above)
Comment (manojpanta) [reply app/assets/stylesheets/nav.scss:17]: text-decoration :none; ?

File: app/helpers/users_helper.rb (hunk 6 above)
Comment (teresa-m-knowles) [reply app/helpers/users_helper.rb:1]: I think this might have gotten created when I scaffolded (is that a verb?) the User model. Def don't need it for now!
